### PR TITLE
SPARC Manual Focus

### PR DESCRIPTION
**Pull request by lazem**

### CODE REVIEW

File: src/odemis/acq/align/autofocus.py [hunk 1]
```diff
@@ -756,6 +756,131 @@ def Sparc2AutoFocus(align_mode, opm, streams=None, start_autofocus=True):
...
+    executeAsyncTask(f, _DoSparc2ManualFocus,
+                     args=(f, turn_on, opm, bl, align_mode, streams))
+    return f
+
+
+def _DoSparc2ManualFocus(future, turn_on, opm, bl, align_mode, streams):
+    """
+    The actual implementation of the manual focus procedure, run asynchronously
+    """
+    try:
```
Comment (Kleijwegt) [inline src/odemis/acq/align/autofocus.py]: You forgot to list the input parameters

File: src/odemis/acq/align/autofocus.py [hunk 2]
```diff
@@ -756,6 +756,131 @@ def Sparc2AutoFocus(align_mode, opm, streams=None, start_autofocus=True):
...
+    f._running_subf = model.InstantaneousFuture()
+    executeAsyncTask(f, _DoSparc2ManualFocus,
+                     args=(f, turn_on, opm, bl, align_mode, streams))
+    return f
+
+
+def _DoSparc2ManualFocus(future, turn_on, opm, bl, align_mode, streams):
+    """
+    The actual implementation of the manual focus procedure, run asynchronously
+    """
```
Comment (Kleijwegt) [inline src/odemis/acq/align/autofocus.py]: You forgot to list the input parameters

File: src/odemis/gui/cont/tabs.py [hunk 3]
```diff
@@ -4025,6 +4180,7 @@ def _onAutofocus(self, active):
                 return
             btn.SetLabel("Auto focus")
 
+
```
Comment (Kleijwegt) [inline src/odemis/gui/cont/tabs.py]: unnecessary new line

File: src/odemis/acq/align/autofocus.py (hunk 2 above)
Comment (lazem) [reply src/odemis/acq/align/autofocus.py]: Thanks, done as requested.

File: src/odemis/acq/align/autofocus.py [hunk 4]
```diff
@@ -756,6 +756,138 @@ def Sparc2AutoFocus(align_mode, opm, streams=None, start_autofocus=True):
...
+            - Change the optical path (closing the slit)
+            - Activate the first focus detector stream (if any)
+        When the manualfocus button is untoggled:
+            - Change the optical path (opening the slit)
+            - Turn off the light
+       turn_on (bool): Toggle the manual focus button on/off
+       align_mode (str): OPM mode, spec-focus or spec-fiber-focus, streak-focus
+       opm: OpticalPathManager object
+       bl: bright light object
+       streams: list of streams
```
Comment (pieleric) [inline src/odemis/acq/align/autofocus.py]: Drop this `streams` argument, as noted below...

File: src/odemis/gui/cont/tabs.py [hunk 5]
```diff
@@ -3918,78 +3932,219 @@ def _onClickFocus(self, evt):
...
+    def _enableManualFocusComponents(self):
+        """
+        Enable/Disable manual focus GUI components (pos slider, detector combobox, grating combobox)
+        Enable/Disable autofocus button
+        Enable/Disable main stream panel
+        """
+        # Ignore streak align mode..
+        align_mode = self.tab_data_model.align_mode.value
+        if align_mode == "streak-align":
+            return
```
[the remark this answers is not in the document]
Comment (lazem) [reply src/odemis/gui/cont/tabs.py]: It's taken care of by `self.panel.pnl_focus.Enable(False)` but this check is still needed since both lens and streak focus go through the same manual focus function. 
Moved it to the caller.

File: src/odemis/gui/main.xrc [hunk 6]
```diff
@@ -2194,43 +2219,62 @@
...
+                    <flag>wxBOTTOM|wxEXPAND|wxALIGN_CENTRE</flag>
+                    <border>10</border>
+                  </object>
+				  <object class="sizeritem">
+				  <object class="wxBoxSizer">
+				  	
+					
+					
+					<object class="sizeritem">
+                        <object class="wxStaticText" name = "cmb_focus_detectors_label">
```
[the remark this answers is not in the document]
Comment (lazem) [reply src/odemis/gui/main.xrc]: Done. Left the gratings label as it's needed in the code.

File: src/odemis/gui/cont/tabs.py [hunk 7]
```diff
@@ -3918,78 +3932,219 @@ def _onClickFocus(self, evt):
...
+        Add/Remove the stream to the current view (updating the StreamTree)
+        Set the focus detectors combobox selection to the the stream's detector
+        :param stream:
+        :param updated:
+        :return:
+        """
+        # Pick the stream to be shown:
+        #  1. Pick the stream which is playing (should_update=True)
+        #  2. Pick the focus stream which is already shown (in v.getStreams())
+        #  3. Pick the first focus stream
```
[the remark this answers is not in the document]
Comment (lazem) [reply src/odemis/gui/cont/tabs.py]: Listening now to the focus stream should_update directly without closure.

File: src/odemis/acq/align/autofocus.py [hunk 8]
```diff
@@ -756,6 +756,135 @@ def Sparc2AutoFocus(align_mode, opm, streams=None, start_autofocus=True):
...
+            future._running_subf.result()
+            future._actions_time.pop(0)
+            future.set_progress(end=time.time() + sum(future._actions_time))
+
+            # Turn on the light
+            if future._manual_focus_state == CANCELLED:
+                logging.info("Manual focus procedure cancelled before the light is on")
+                raise CancelledError()
+            bl.power.value = bl.power.range[1]
+            bl.emissions.value = [1] * len(bl.emissions.value)
```
Comment (pieleric) [inline src/odemis/acq/align/autofocus.py]: Actually, you don't need to wait for the path to be ready before turning on the light. You can do it at the same time. So in-between `opm.setPath(align_mode)` and `.result()`. In this case, you only have one step (of 10s).

File: src/odemis/acq/align/autofocus.py [hunk 9]
```diff
@@ -756,6 +756,135 @@ def Sparc2AutoFocus(align_mode, opm, streams=None, start_autofocus=True):
...
+    with future._manual_focus_lock:
+        if future._manual_focus_state == FINISHED:
+            return False
+        future._manual_focus_state = CANCELLED
+        future._running_subf.cancel()
+        logging.debug("Sparc2AManualFocus cancellation requested.")
+
+    return True
+
+def Sparc2AManualFocus(turn_on, align_mode, opm, bl):
```
Comment (pieleric) [inline src/odemis/acq/align/autofocus.py]: Let's drop the `A`, -> `Sparc2ManualFocus()`

**Comment by pieleric** [comment]
> I have somewhat the feeling that some corner cases might not be properly handled (eg, cancel manual focus on streak cam). Admittedly that's quite a bit of a pain to test everything (manually). Make sure to be extra careful.

Sorry, I realise that actually when switching to manual focus, it's not really possible to cancel, unless changing mode or tab. So it might be a little bit easier to handle.

File: src/odemis/acq/align/autofocus.py (hunk 8 above)
Comment (lazem) [reply src/odemis/acq/align/autofocus.py]: Done with the suggested refactoring. Just kept the canceller so at least the progress bar appears stopped (even if the opm future didn't get cancelled).

File: src/odemis/gui/cont/tabs.py [hunk 10]
```diff
@@ -3202,6 +3204,7 @@ def __init__(self, name, button, panel, main_frame, main_data):
         self._spot_stream = spot_stream
 
         self._ccd_stream = None
+        self._ccd_spe = None
```
Comment (Kleijwegt) [inline src/odemis/gui/cont/tabs.py]: What does SPE stand for?

File: src/odemis/gui/cont/tabs.py [hunk 11]
```diff
@@ -3313,6 +3286,36 @@ def __init__(self, name, button, panel, main_frame, main_data):
         else:
             self.panel.pnl_fib_focus.Show(False)
 
+        # Manual focus mode initialization
+        # All the `blue` streams one for each detector to adjust for focus
```
Comment (Kleijwegt) [inline src/odemis/gui/cont/tabs.py]: Seems like in this sentence a typo slipped in? Maybe a comma need to be add? Maybe the word All needs to be Add? What do you mean exactly?

File: src/odemis/gui/cont/tabs.py [hunk 12]
```diff
@@ -3918,78 +3922,188 @@ def _onClickFocus(self, evt):
...
+                d,
+                d.data,
+                emitter=None,  # actually, the brightlight, but the autofocus procedure takes care of it
+                focuser=focuser,
+                detvas=get_local_vas(d, hw_settings_config),
+                forcemd={model.MD_POS: (0, 0),
+                         model.MD_ROTATION: 0},
+                acq_type=model.MD_AT_SLIT,
+            )
+            speclines.tint.value = (0, 64, 255)  # colour it blue
```
Comment (Kleijwegt) [inline src/odemis/gui/cont/tabs.py]: Maybe the imported blue should be a color defined in 

> src/odemis/gui/__init__.py

So maybe import from this place or reference to it in a comment

File: src/odemis/gui/cont/tabs.py [hunk 13]
```diff
@@ -3918,78 +3922,188 @@ def _onClickFocus(self, evt):
...
+        Play the stream associated with the chosen detector
+        """
+        # Find the stream related to this selected detector
+        idx = self.panel.cmb_focus_detectors.GetSelection()
+        stream = self._focus_streams[idx]
+        # Play this stream (set both should_update and is_active with True)
+        stream.should_update.value = True
+        stream.is_active.value = True
+
+    def add_combobox_control(self, label_text, value=None, conf=None):
```
Comment (Kleijwegt) [inline src/odemis/gui/cont/tabs.py]: Maybe this method should be private?

File: src/odemis/acq/align/autofocus.py [hunk 14]
```diff
@@ -757,6 +757,82 @@ def Sparc2AutoFocus(align_mode, opm, streams=None, start_autofocus=True):
     return f
 
 
+def _cancelSparc2ManualFocus(future):
```
[the remark this answers is not in the document]
Comment (lazem) [reply src/odemis/acq/align/autofocus.py]: All methods for focus are inside autofocus.py, the public ones are `Sparc2AutoFocus` and `Sparc2ManualFocusused`, used by `Sparc2AlignTab` object. The rest is private.

File: src/odemis/gui/cont/tabs.py (hunk 10 above)
Comment (lazem) [reply src/odemis/gui/cont/tabs.py]: Good question! This is a stream controller for the ccd stream, it's initialized in here 

```
self._ccd_spe = self._stream_controller.addStream(ccd_stream,
                                add_to_view=self.panel.vp_align_lens.view)
```

Maybe @pieleric can tell us, I used it as it's defined.

File: src/odemis/gui/cont/tabs.py (hunk 13 above)
Comment (lazem) [reply src/odemis/gui/cont/tabs.py]: It needs to be the same signature to be used correctly by `create_axis_entry`.

File: src/odemis/gui/cont/tabs.py (hunk 12 above)
Comment (lazem) [reply src/odemis/gui/cont/tabs.py]: Done as requested.

File: src/odemis/gui/cont/tabs.py (hunk 10 above)
Comment (Kleijwegt) [reply src/odemis/gui/cont/tabs.py]: Maybe it is nice to add a comment with the explanation behind the initialization of this variable?

Comment (pieleric) [reply src/odemis/gui/cont/tabs.py]: Stream Panel Entry.

File: src/odemis/gui/cont/tabs.py [hunk 15]
```diff
@@ -3918,78 +3922,188 @@ def _onClickFocus(self, evt):
...
-            #     # Enable autofocus if manual focus btn not pushed
-            #     self.panel.btn_autofocus.Enable(True)
-            #     self.panel.gauge_autofocus.Enable(True)
 
+            self._mf_future = Sparc2ManualFocus(main.opm, bl, opath, mf_toggled=False)
+            self._mf_future.add_done_callback(self._onManualFocusFinished)
+            # Update GUI
+            self._pfc_manual_focus = ProgressiveFutureConnector(self._mf_future, gauge)
+
+    def _onManualFocusReady(self, future):
```
Comment (pieleric) [inline src/odemis/gui/cont/tabs.py]: You should protect this function with a `@call_in_wx_main`, as it will be run in a separate thread and it accesses the GUI.

File: src/odemis/gui/cont/tabs.py [hunk 16]
```diff
@@ -4003,6 +4117,8 @@ def _onAutofocus(self, active):
 
             # GUI stream bar controller pauses the stream
             btn.SetLabel("Cancel")
+            self.panel.btn_manual_focus.Enable(False)
+            self._enableFocusComponents(ccd_stream=False)
```
Comment (pieleric) [inline src/odemis/gui/cont/tabs.py]: Please explicitly write `manual=False`

File: src/odemis/gui/cont/tabs.py [hunk 17]
```diff
@@ -3918,78 +3922,188 @@ def _onClickFocus(self, evt):
...
+        Called when a focus stream "should_update" state changes
+        Add/Remove the stream to the current view (updating the StreamTree)
+        Set the focus detectors combobox selection to the the stream's detector
+        # Pick the stream to be shown:
+        #  1. Pick the stream which is playing (should_update=True)
+        #  2. Pick the focus stream which is already shown (in v.getStreams())
+        #  3. Pick the first focus stream
+        """
+        focusedview = self.tab_data_model.focussedView.value
+        should_update_stream = next((s for s in self._focus_streams if s.should_update.value == True), None)
```
Comment (pieleric) [inline src/odemis/gui/cont/tabs.py]: Drop the `== True`. Booleans can be directly used as-is.

File: src/odemis/gui/cont/tabs.py [hunk 18]
```diff
@@ -3918,78 +3922,188 @@ def _onClickFocus(self, evt):
...
+        if should_update_stream:
+            # This stream should be shown, remove the other ones first
+            for st in focusedview.getStreams():
+                if st in self._focus_streams and st is not should_update_stream:
+                    focusedview.removeStream(st)
+            focusedview.addStream(should_update_stream)
+            # Set the focus detectors combobox selection
+            try:
+                istream = self._focus_streams.index(should_update_stream)
+                if istream != -1:
```
Comment (pieleric) [inline src/odemis/gui/cont/tabs.py]: Drop this `if`. `index()` always returns >= 0.

File: src/odemis/gui/cont/tabs.py (hunk 10 above)
Comment (lazem) [reply src/odemis/gui/cont/tabs.py]: Added a comment for the reason of keeping it as attribute

File: src/odemis/acq/align/autofocus.py [hunk 19]
```diff
@@ -757,6 +757,82 @@ def Sparc2AutoFocus(align_mode, opm, streams=None, start_autofocus=True):
...
+    logging.debug("Cancelling manual focus...")
+
+    with future._manual_focus_lock:
+        if future._state == FINISHED:
+            return False
+        future._state = CANCELLED
+        future._running_subf.cancel()
+        logging.debug("Sparc2AManualFocus cancellation requested.")
+
+    return True
```
[the remark this answers is not in the document]
Comment (lazem) [reply src/odemis/acq/align/autofocus.py]: That's right, it keeps going when the return is not True.

File: src/odemis/gui/cont/tabs.py [hunk 20]
```diff
@@ -3918,78 +3922,188 @@ def _onClickFocus(self, evt):
...
+        stream.is_active.value = True
+
+    def add_combobox_control(self, label_text, value=None, conf=None):
+        """ Add a combo box to the focus panel
+        # HACK to be used to automatically bind gratings to the combobox in the focus panel
+        """
+        # No need to create components, defined already on the xrc file
+        return self.panel.cmb_focus_gratings_label, self.panel.cmb_focus_gratings
+
+    def _enableFocusComponents(self, manual=False, ccd_stream=False):
```
[the remark this answers is not in the document]
Comment (lazem) [reply src/odemis/gui/cont/tabs.py]: Dropped default value of `manual`.

File: src/odemis/gui/cont/tabs.py [hunk 21]
```diff
@@ -3918,78 +3924,188 @@ def _onClickFocus(self, evt):
...
+            if model.hasVA(speclines, "detReadoutRate"):
+                try:
+                    speclines.detReadoutRate.value = speclines.detReadoutRate.range[1]
+                except AttributeError:
+                    speclines.detReadoutRate.value = max(speclines.detReadoutRate.choices)
+            focus_streams.append(speclines)
+
+        return focus_streams
+
+    def _ensureOneFocusStream(self, _):
```
Comment (pieleric) [inline src/odemis/gui/cont/tabs.py]: Should be also protected by `@call_in_wx_main`.

**Comment by lazem** [comment]
> Looks fine. There is just one extra change that I think is needed.
> Can you add the suggested line, and squash the commits?

Added the suggested line and squashed the commits. 
@pieleric
